sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1825,17 +1830,27 @@ pi_result piDevicePartition(pi_device Device,
   }
 
   try {
+    pi_platform Platform = Device->Platform;
     auto ZeSubdevices = new ze_device_handle_t[Count];
     ZE_CALL(zeDeviceGetSubDevices(Device->ZeDevice, &Count, ZeSubdevices));
 
     // Wrap the Level Zero sub-devices into PI sub-devices, and write them out.
     for (uint32_t I = 0; I < Count; ++I) {
-      OutDevices[I] = new _pi_device(ZeSubdevices[I], Device->Platform,
-                                     true /* isSubDevice */);
-      pi_result Result = OutDevices[I]->initialize();
-      if (Result != PI_SUCCESS) {
-        delete[] ZeSubdevices;
-        return Result;
+      pi_device Dev = Platform->getDeviceFromNativeHandle(ZeSubdevices[I]);
+      if (Dev) {
+        OutDevices[I] = Dev;
+      } else {
+        std::unique_ptr<_pi_device> PiSubDevice(
+            new _pi_device(ZeSubdevices[I], Platform));
```

Comment:
I understand this is not a new problem, but who will delete these? I think you should call piDeviceRelease of sub-devices in piDeviceRelease of root-device

Reply:
Please check/add

Reply:
Yes, I verified with valgrind, and they are released properly. No memory leak is detected for sub-devices.
Anyway, if SYCL RT does not call piDeviceRelease for pi_device returned from the plugin, it is a bug in SYCL RT.

Reply:
But if SYCL RT **does** call `piDeviceRelease ` on every PI device returned from `piDevicePartition` then your changes will break it: multiple partitioning calls will now return same PI devices, and so they will be deleted too early, at the first `piDeviceRelease` of a sub-device.

Reply:
I see. 
Then, I will call piDeviceRetain() when we reuse the same pi_device for the same sub-device.

Reply:
Yes, and add a test covering this.

Reply:
okay. I will extend the existing test https://github.com/intel/llvm-test-suite/blob/intel/SYCL/Basic/query.cpp to test reusing the same sub-device.
It is going to be a separate PR in that repo later.